flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java

```diff
@@ -0,0 +1,167 @@
+ * The owner is responsible for the writers' lifecycle from the moment of creation.
+ *
+ * <p>Partitions are released in the following cases:
+ * <ol>
+ *     <li>{@link ResultPartitionWriter#fail(Throwable)} and {@link ResultPartitionWriter#close()} are called
+ *     if the production has failed.
+ *     </li>
+ *     <li>{@link ResultPartitionWriter#finish()} and {@link ResultPartitionWriter#close()} are called
+ *     if the production is done. The actual release can take some time depending on implementation details,
+ *     e.g. if the `end of consumption' confirmation from the consumer is being awaited implicitly
+ *     or the partition is later released by {@link ShuffleEnvironment#releasePartitions(Collection)}.</li>
+ *     <li>{@link ShuffleEnvironment#releasePartitions(Collection)} is called outside of the producer thread,
+ *     e.g. to manage the lifecycle of BLOCKING result partitions which can outlive their producers.</li>
+ * </ol>
+ * The partitions, which currently still occupy local resources, can be queried with
+ * {@link ShuffleEnvironment#getPartitionsOccupyingLocalResources}.
+ *
+ * <h2>Input gate management.</h2>
+ *
+ * <p>The interface implements a factory for the input gates: {@link ShuffleEnvironment#createInputGates}.
+ * The created gates are grouped per owner. The owner is responsible for the gates' lifecycle from the moment of creation.
+ *
+ * <p>When the input gates are created, it can happen that not all consumed partitions are known at that moment
+ * e.g. because their producers have not been started yet. Therefore, the {@link ShuffleEnvironment} provides
+ * a method {@link ShuffleEnvironment#updatePartitionInfo} to update them externally, when the producer becomes known.
+ * The update mechanism has to be threadsafe because the updated gate can be read concurrently from a different thread.
+ *
+ * @param <P> type of provided result partition writers
+ * @param <G> type of provided input gates
+ */
+public interface ShuffleEnvironment<P extends ResultPartitionWriter, G extends InputGate> extends AutoCloseable {
+
+	/**
+	 * Start the internal related services before using the shuffle service environment.
+	 *
+	 * @return a port to connect for the shuffle data exchange, -1 if only local connection is possible.
+	 */
+	int start() throws IOException;
+
+	/**
+	 * Create a context of the shuffle input/output owner used to create partitions or gates belonging to the owner.
+	 *
+	 * <p>This method has to be called only once to avoid duplicated internal metric group registration.
+	 *
+	 * @param ownerName the owner name, used for logs
+	 * @param executionAttemptID execution attempt id of the producer or consumer
+	 * @param parentGroup parent of shuffle specific metric group
+	 * @return context of the shuffle input/output owner used to create partitions or gates belonging to the owner
+	 */
+	ShuffleIOOwnerContext createShuffleIOOwnerContext(
```

Comment:
it would be useful the partition lifecycle management if this method would also accept the JobID..

Reply:
Shuffle service does not need it at the moment, would be nice actually to keep it decoupled from the job id, I would suggest we discuss it at the corresponding partition lifecycle PR and add it if needed.

Reply:
see this discussion then: https://github.com/apache/flink/pull/8687#discussion_r292370552